### PR TITLE
kan-268 added outfilter to redux

### PR DESCRIPTION
**Pull request by jeanawhou**
Added outfilter to redux
Allow the user to filter by more than one plotline
Allow user to export filtered outline

PS: I picked this ticket to learn more how plottr manages the state. I would love to know comments guys 
ticket link: https://app.ora.pm/p/4c35de9a18594ccbaee3377dcc546780?v=0&t=k&c=cc22f6e6d0ed4040a832aa0c7c78449a
Preview on slack #dev-desktop

### CODE REVIEW

File: lib/pltr/v2/helpers/cards.js
Comment (Quiescent) [inline lib/pltr/v2/helpers/cards.js]: I've picked up on two issues here:

1. The check on 28 seems redundant.  You should be able to delete it safely.
2. I think that `&& !!currentLine.length` might be redundant.
    I also feel like the extra check makes this harder to read.
    If we already know that it's an array, then whether it has a length is already checked in `.includes()`.

i.e. I think that you can rewrite 28-29 as:
```javascript
  if (Array.isArray(currentLine)) return !currentLine.includes(line.id)
```

File: lib/pltr/v2/helpers/cards.js
Comment (Quiescent) [inline lib/pltr/v2/helpers/cards.js]: If `currentLine` is falsy, then there's no need to check whether it's also not an array.

File: lib/pltr/v2/helpers/cards.js
Comment (Quiescent) [inline lib/pltr/v2/helpers/cards.js]: You can remove the `Array.isArray` from line 31 and line 33 safely.
Then this line looks much more straight forward.

File: src/app/components/outline/OutlineView.js
Comment (Quiescent) [inline src/app/components/outline/OutlineView.js]: There's no need to make this async.
We're not calling out to any APIs, and the javascript interpreter loop must finish before we do a redraw.

File: src/app/components/outline/OutlineView.js
Comment (Quiescent) [inline src/app/components/outline/OutlineView.js]: Stylistically, this is too much logic in the component.
It would be best to move this to the actions file or to the reducer (I'm leaning to the second).
The choice between the reducer and actions is made based on whether the logic is async or not.
If it's async, it goes into actions (and we'll need to add an async-aware middleware).
If it's not async (which I believe to be the case), then the reducer is the correct home for this logic.

File: src/app/components/outline/OutlineView.js
Comment (Quiescent) [inline src/app/components/outline/OutlineView.js:49]: It seems that your change relies on `outlineFilter` on `ui` rather than on the state.
(I don't see us calling `setState` to change `currentLine` anywhere.)
If I'm right about that, then it should probably be deleted (along with the other references to it later on).

File: src/app/components/outline/OutlineView.js
Comment (Quiescent) [inline src/app/components/outline/OutlineView.js]: I think this is an error?
We never set `currentLine` anymore so surely we should delete it?

File: src/app/components/outline/OutlineView.js
Comment (Quiescent) [inline src/app/components/outline/OutlineView.js]: Same thing here re. `currentLine`.

File: src/app/components/outline/OutlineView.js
Comment (Quiescent) [inline src/app/components/outline/OutlineView.js]: Another reference to `currentLine`.

File: src/app/components/outline/OutlineView.js
Comment (Quiescent) [inline src/app/components/outline/OutlineView.js]: Same thing re. `currentLine` here.

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js]: Please use `const` instead of `let` as much as you can.
`let` makes the code harder to read because the reader must search for places where `outlineFilter` could be re-assigned.

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js]: I think that this should stay `const`.
I don't see a reassignment anywhere.
Am I missing something?

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js]: Won't this filter out any paragraphs if there's an active filter?

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js]: I think that this `let` should be a `const`.

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js]: This adds an `undefined` to the list.
Did you mean: `return []`?

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js]: This side effect shouldn't happen inside of a call to `flatMap`.
Am I right in guessing that this would do the trick:
```
sortedCards.forEach((c) => {
  if (outlineFilter && !outlineFilter.includes(c.lineId))  return
  if (outlineFilter) paragraphs.push(new Paragraph({ text: title, heading: HeadingLevel.HEADING_2 }))
})

let cardParagraphs = sortedCards.flatMap...
```

**Review by Quiescent — CHANGES_REQUESTED** [review]
Great approach.
I've picked up a few issues regarding correctness in the export and style in the component.

**Comment by jeanawhou** [comment]
@Quiescent thank you for the very informative review!

**Review by cameronsutter — COMMENTED** [review]
Nothing to add on top of Ed's comments

**Review by murfdog64 — COMMENTED** [review]
I'm a little late, nothing to add 😅

**Comment by jeanawhou** [comment]
@Quiescent i just updated this one. There are few conditions that i wasn't able to follow because it didn't work on my end especially this one https://github.com/cameronsutter/plottr_electron/pull/579#discussion_r609737774 i need help with this one. I pushed a working code though

**Comment by Quiescent** [comment]
> @Quiescent i just updated this one. There are few conditions that i wasn't able to follow because it didn't work on my end especially this one [#579 (comment)](https://github.com/cameronsutter/plottr_electron/pull/579#discussion_r609737774) i need help with this one. I pushed a working code though

I'll take a look at the current version. :)

File: lib/pltr/v2/helpers/cards.js
Comment (Quiescent) [inline lib/pltr/v2/helpers/cards.js]: I might've missed something here.  Why do you return `!currentLine` instead of `false`?

File: lib/pltr/v2/helpers/cards.js
Comment (Quiescent) [inline lib/pltr/v2/helpers/cards.js]: This should be `else`, and please use `!==` instead of `!=` to avoid unintentional type casts.

File: lib/pltr/v2/reducers/ui.js
Comment (Quiescent) [inline lib/pltr/v2/reducers/ui.js]: This can be `!action.filter` it covers `null` because `null` is falsey.

File: lib/pltr/v2/reducers/ui.js
Comment (Quiescent) [inline lib/pltr/v2/reducers/ui.js]: We can simplify this to `Array.isArray(state.outlineFilter)` i.e. delete the `!!state.outlineFilter` because `Array.isArray` already checks for null and undefined.

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js]: It seems to me that we want to add a level two-paragraph when we have at least one card that wasn't filtered out(?)
This code as it stands is trying to do too many things at the same time.

This has resulted in a bug: if you return from the body of a call to `flatMap` with undefined (line 59) then you'll get an undefined in an array (e.g. `[x, y, ...undefined, ..., z]`)

If we instead separate things then the problem becomes simpler.
Let's start by getting a collection of cards that aren't filtered out,
```javascript
  const filteredCards = !outlineFilter ? sortedCards : sortedCards.filter((card) => {
    outlineFilter.includes(card.lineId)
  }
```
then let's add the level 2 paragraph if there are cards,

```javascript
  if (filteredCards.length) {
    paragraphs.push(new Paragraph({ text: title, heading: HeadingLevel.HEADING_2 }))
  }
```

finally, let's add the paragraphs for each card.
```javascript
  const cardParagraphs = filteredCards.flatMap((c) =>
    card(c, linesById, namesMapping, customAttrs, doc, options)
  )
```

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [reply src/common/exporter/word/exporters/outline.js]: Forget what I said before.  I see how you handled this later on.

There is a problem here.  If we have an outline filter, but `options.outline.sceneCards` is `false`, we won't end up with a level 2 paragraph (whereas before your change we used to always have a level 2 paragraph if we hit this line of code).

I think that the check should be `if (!outlineFilter || !options.outline.sceneCards)`.
WDYT?

File: lib/pltr/v2/helpers/cards.js
Comment (Quiescent) [inline lib/pltr/v2/helpers/cards.js]: The length check is redundant here.

**Review by Quiescent — COMMENTED** [review]
I've gone over the PR again as it stands.
There are still a few places where the logic could be simpler and clearer.
There's also a longer (and hopefully correct this time!) response to the exporter problem.

**Comment by jeanawhou** [comment]
@Quiescent i have removed the unnecessary checks, i dont know why they didn't work last time i remove them and also i hope i got the filter right. i think lodash's includes is what i missed last time. Thanks again!

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js]: This flag seems to be redundant.  It's reset to zero whenever we enter this block anyway.
Please could you remove it?

File: lib/pltr/v2/helpers/cards.js
Comment (Quiescent) [inline lib/pltr/v2/helpers/cards.js]: This line is redundant.  If it's an array then we don't need to check whether it has a length to run `.includes` on it.

File: lib/pltr/v2/helpers/cards.js
Comment (Quiescent) [reply lib/pltr/v2/helpers/cards.js]: I think that this line should instead read:
```
else return line.id !== currentLine
```
because we know at this point that `currentLine` must be truthy.

File: lib/pltr/v2/reducers/ui.js
Comment (Quiescent) [inline lib/pltr/v2/reducers/ui.js]: `!state.outlineFilter` is redundant here.
We can instead just write:
```
else if (!Array.isArray(state.outlineFilter))
```
because it'll already check whether it's `null`, `undefined` or `false`.

**Review by Quiescent — COMMENTED** [review]
@jeanawhou I've found few places where I still see problems.  I hope that you find this helpful despite how long it's taking to work through these problems.
If you'd like to chat live about the changes, perhaps we can schedule some time to talk.

**Comment by Quiescent** [comment]
@jeanawhou I'm going to resolve all of the issues from previous reviews to reduce noise.
You can look at the latest review for the changes that I think need to happen before merging this.

File: src/common/exporter/word/exporters/outline.js
Comment (Quiescent) [inline src/common/exporter/word/exporters/outline.js:47]: I think that there's a slight problem here.
In the case that we're not exporting scene cards, we probably still want this level 2 heading.
So can we add a check here like so:
```
if (!options.outline.sceneCards) {
  paragraphs.push(new Paragraph({ text: title, heading: HeadingLevel.HEADING_2 }))
}
```

Comment (Quiescent) [reply src/common/exporter/word/exporters/outline.js:47]: @cameronsutter, could you please clarify something about this change.

I've suggested that we now need to check whether we're not exporting scene cards and if that's the case, add a level two heading.
The check would maintain what the code did before.

Bug @jeanawhou has tested the change and isn't sure what a level 2 heading looks like.

I don't want to hold up the PR if this isn't an important detail.
I'm also struggling to justify why it should remain as it was because I can't provide a good test case to see the change.

A possible test might be to check out staging and export with scene cards turned off and then switch to this branch and export again (with scene cards off) to see the difference.
WDYT?

Comment (Quiescent) [reply src/common/exporter/word/exporters/outline.js:47]: I discussed and made the change with @jeanawhou.

**Review by Quiescent — APPROVED** [review]
Looks good.
LGTM!

**Pull request merged**
